Userland/Libraries/LibWeb/HTML/Worker.cpp

```diff
@@ -0,0 +1,363 @@
+    if (!url.is_valid()) {
+        dbgln_if(WEB_WORKER_DEBUG, "WebWorker: Invalid URL loaded '{}'.", script_url);
+        return DOM::SyntaxError::create("Failed to parse selector");
+    }
+
+    // 5. Let worker URL be the resulting URL record.
+    auto worker_url = url;
+
+    // 6. Let worker be a new Worker object.
+    auto worker = adopt_ref(*new Worker(script_url, move(options), document));
+
+    // 7. Let outside port be a new MessagePort in outside settings's Realm.
+    auto outside_port = MessagePort::create();
+
+    // 8. Associate the outside port with worker->
+    worker->m_outside_port = outside_port;
+
+    // 9. Run this step in parallel:
+    worker->run_a_worker(worker_url, outside_settings, outside_port, options);
+
+    // 10. Return worker->
+    return worker;
+}
+
+// https://html.spec.whatwg.org/multipage/workers.html#run-a-worker
+void Worker::run_a_worker(AK::URL& url, EnvironmentSettingsObject& outside_settings, MessagePort& outside_port, WorkerOptions const& options)
+{
+    // 1. Let is shared be true if worker is a SharedWorker object, and false otherwise.
+    // FIXME: SharedWorker support
+    auto is_shared = false;
+    auto is_top_level = false;
+
+    // 2. Let owner be the relevant owner to add given outside settings.
+    // FIXME: Support WorkerGlobalScope options
+    if (!is<HTML::WindowEnvironmentSettingsObject>(outside_settings))
+        TODO();
+
+    //auto owner = m_document;//outside_settings.responsible_document();
+
+    // 3. Let parent worker global scope be null.
+    // 4. If owner is a WorkerGlobalScope object (i.e., we are creating a nested dedicated worker),
+    //    then set parent worker global scope to owner.
+    // FIXME: Support for nested workers.
+
+    // 5. Let unsafeWorkerCreationTime be the unsafe shared current time.
+    //struct timeval unsafe_worker_creation_time; //Currently unused
+
+    // 6. Let agent be the result of obtaining a dedicated/shared worker agent given outside settings
+    // and is shared. Run the rest of these steps in that agent.
+    // FIXME: Agent support
+
+    // 7. Let realm execution context be the result of creating a new JavaScript realm given agent and the following customizations:
+    m_worker_realm = JS::Realm::create(m_worker_vm);
+
+    //      7a. For the global object, if is shared is true, create a new SharedWorkerGlobalScope object.
+    //      7b. Otherwise, create a new DedicatedWorkerGlobalScope object.
+    // FIXME: Proper support for both SharedWorkerGlobalScope and DedicatedWorkerGlobalScope
+    if (is_shared)
+        TODO();
+
+    // FIXME: Make and use subclasses of WorkerGlobalScope, however this requries JS::GlobalObject to
+    //        play nicely with the IDL interpreter, to make spec-compliant extensions, which it currently does not.
+    m_worker_scope = static_cast<JS::GlobalObject*>(m_worker_vm->heap().allocate_without_global_object<JS::GlobalObject>());
```

Comment:
Why are you `static_cast`ing here? It should just return that type already

Reply:
Leftover from when I was experimenting with using the `GlobalWorkerScope` class for the main scope. That didn't work (I think) because of issues with the IDL generator working with the `JS::GlobalObject` class. I could certainly use some input on this problem, but that's probably a separate discussion. I'll clean it up.